# Assembly
Stitching the error corrected reads to prepare continuous long stretches of sequence.

The aim was to assemble the genome using long reads. Hence, [Flye](https://github.com/fenderglass/Flye) was used to assemble the contigs from the read files. Alternatively [Canu](https://github.com/marbl/canu) can also be used for long read based assembly.
```bash
# Running flye assembler
flye --nano-raw ~/path/to/dir/in/lr.corrected.fa.gz -o ~/path/to/dir/out -t $threads
```
> The output of this assembly is `contigs.fa`.

The resulting contigs were furhter stitched together to prepare the extended contigs with the help of evidences from the corrected long reads. We used [LongStitch](https://github.com/bcgsc/LongStitch) for the purpose.
```bash
# Stiching the contigs
longstitch ntLink-arks draft=~/path/to/dir/in/contigs.fa reads=~/path/to/dir/in/lr.corrected.fa.gz \
 out_prefix=$prefix t=$threads G=1e9
```
> The output of this step is `long.contigs.fa`.

Next, the polishing and consensus calling of the long contigs were performed with the help of the error corrected long reads using [medaka](https://github.com/nanoporetech/medaka).
```bash
# Running medaka
medaka_consensus -i ~/path/to/dir/in/lr.corrected.fa.gz -d ~/path/to/dir/in/long.contigs.fa \
 -o ~/path/to/dir/out -m r104_e81_hac_g5015 -t $threads
```
> The appropriate model assignment through the option `-m` can be cheked by using `medaka tools list\_models`.  
> The output of the consensus call is `consensus.contigs.fa`.

Subsequent polishing was performed using with short reads by [Polca](https://github.com/alekseyzimin/masurca), a program distributed as a part of MaSuRCA genome assembly toolkit.
```bash
# Running polca
polca.sh -a ~/path/to/dir/in/consensus.contigs.fa -r '~/path/to/dir/in/sr.corrected.R1.fq.gz \
 ~/path/to/dir/in/sr.corrected_R2.fq.gz' -t $threads -m 2G
 ```
> The output of this polishing step is `consensus.pc.contigs.fa`.

Subsequently, the polca corrected contigs were screened for contaminants using [kraken2](https://github.com/DerrickWood/kraken2) and the identified contaminants were removed using [KrakenTools](https://github.com/jenniferlu717/KrakenTools).
```bash
# Running kraken
kraken2 --db $database --threads $threads --output ~/path/to/dir/out/kraken.out --confidence 0.10 \
 --report ~/path/to/dir/out/report.kraken <(dustmasker -in ~/path/to/dir/in/consensus.pc.contigs.fa \
 -outfmt fasta | sed -e '/^>/!s/[atgc]/N/g')

# Extrcating an identified contaminant
krakentools/extract_kraken_reads.py -k ~/path/to/dir/in/kraken.out -s ~/path/to/dir/in/consensus.pc.contigs.fa \
 -t 953 -r ~/path/to/dir/in/report.kraken --include-children -o ~/path/to/dir/out/prok.contigs.fa

# Isolating the draft assembly using an in-house python file
python ~/path/to/dir/in/filter_assembly.py
```
> As a result, the draft assembly `draft.assembly.fa` was recovered.

Subsequently, the mitochondrial contigs within the draft assembly were identified using [GetOrganelle](https://github.com/Kinggerm/GetOrganelle).
```bash
# Extracting and annotating mito genome 
get_organelle_from_assembly.py -g $flye.graph -F animal_mt -o ~/path/to/dir/out --min-depth 10 -t $threads
seqkit replace -p '.*' -r $prefix -o ~/path/to/dir/out/mito.draft.fa ~/path/to/dir/in/*.fa
```
> Only one contig in the draft assembly `draft.assembly.fa` was recovered as the mitocondrial fragment `mito.draft.fa`.

The draft assemblies were further assessed as mentioned in the post-assembly and mitogenome sections.
