# Mitogenome & Numt
Mitochondrial genome assembly, annotation and NUMT identification.

As only one continuous fragment was detected as a mitochondrial contig during the assembly step, this fragment was used to retrieve putative mitochondrial reads. To this end, the error corrected long reads were first mapped 
to the draft assembly using [bwa](https://github.com/lh3/bwa) aligner.
```bash
# Genome-wide mapping
bwa mem -t $threads ~/path/to/dir/in/draft.assembly.fa ~/path/to/dir/in/lr.corrected.fa.gz | samtools view -buS - \
 | samtools sort - -o ~/path/to/dir/out/lr.draft.sorted.bam
```
> The output of this assembly is the genome-wide mapping file `lr.draft.sorted.bam`.

From this genome-wide mapping file, the reads mapping to the identified mitochondrial contig were recovered.
```bash
# Stiching the contigs
samtools view -h ~/path/to/dir/in/lr.draft.sorted.bam $mitocontigID:position > mitoread.bam
samtools sort mitoread.bam > mitoread.sorted.bam
samtools index mitoread.sorted.bam 
samtools fastq mitoread.sorted.bam > mitoreads.fa
```
> The output of this step is the mitochondria specific long read file `mitoreads.fasta`.

Next, the mitochondria specifc reads were assembled with [flye](https://github.com/fenderglass/Flye) under `-meta` option.
```bash
# Running flye
flye --meta --nano-corr ~/path/to/dir/in/mitoreads.fa --out-dir ~/path/to/dir/out/mitochondrion --threads $threads
```
> The output of the assembly is `mitochondrion.fa`.

Subsequently, the assembled mitochondrial genome was annotated with the online server [Mitos2](http://mitos2.bioinf.uni-leipzig.de/index.py). Annotation of the low complexity control region was conducted using the online tool 
[TRF](https://tandem.bu.edu/trf/trf.html) for tandem repeats and [IUPACpal](https://github.com/steven31415/IUPACpal) for palindromes.
```bash
# Running IUPACpal
./IUPACpal -f ~/path/to/dir/in/mitochondrion.fa
 ```
To identify potential insertions of mitochondrial fragments within the nuclear genome (NUMTs), the nuclear genome was explored for mitochondrial signature using [BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi) search. The resulting BLAST output was filtered with the help of an in-house script (script/blastn_filter_numt.py) to retrieve the potential NUMTs.
```bash
blastn -query ~/path/to/dir/in/mitochondrion.fa -db ~/path/to/dir/in/draft.assembly.fa -outfmt \
 "7 qaccver saccver sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore qcovs qcovhsp" \
 -evalue 0.00005 -task blastn > blastn_raw.txt

# Filter the blast search output
python blastn_filter_numt.py blastn_raw.txt
```
