# Repeatlibrary
Identification and annotation of repeats in the draft genome

As an essential step prior to the de-novo genome annotation, we identified and annotated different classes of repeats in the draft genome. The process of repeat annotation
is a never ending task. However, to prepare a comprehensive repeat landscape of the draft genome, we adopted recently benchmarked pipelines that brings a brilliant compilation
of analytical tools and publicly available databases under a common umbrella.

We used **The Extensive *de novo* TE Annotator** [EDTA](https://github.com/oushujun/EDTA). EDTA appears as an one stop solution for repeat library preparation.
```bash
# Running EDTA
EDTA.pl --genome /path/to/dir/draft.assembly.fa --overwrite 1 --sensitive 1 --anno 1 --evaluate 1 --threads 110
```
The output from EDTA can be parsed to RepeatMasker to mask the draft genome which will further be analysed for genome annotation.

Next, we tried **TransposonUltimate** [TransposonUltimate](https://github.com/DerKevinRiehl/TransposonUltimate). Here, we particularly worked with the bundle *Transposon Annotator "reasonaTE"*. After successful installation of the required tools within this bundle, we executed each program separately to perfform parallel data analyses.
```bash
# Making the project working directory
mkdir /home/sahoork/beetle/analysis/mRNA/TEspace
reasonaTE -mode createProject -projectFolder /path/to/dir -projectName $dirname -inputFasta /path/to/dir/in/draft.assembly.fa

# Running each programme separately
reasonaTE -mode annotate -projectFolder /path/to/dir -projectName $dirname -tool helitronScanner
reasonaTE -mode annotate -projectFolder /path/to/dir -projectName $dirname -tool ltrHarvest 
reasonaTE -mode annotate -projectFolder /path/to/dir -projectName $dirname -tool mitefind
reasonaTE -mode annotate -projectFolder /path/to/dir -projectName $dirname -tool mitetracker xxxxx -w 10
reasonaTE -mode annotate -projectFolder /path/to/dir -projectName $dirname -tool must
reasonaTE -mode annotate -projectFolder /path/to/dir -projectName $dirname -tool repeatmodel 
reasonaTE -mode annotate -projectFolder /path/to/dir -projectName $dirname -tool repMasker
reasonaTE -mode annotate -projectFolder /path/to/dir -projectName $dirname -tool sinefind
reasonaTE -mode annotate -projectFolder /path/to/dir -projectName $dirname -tool sinescan xxxxx -k 10
reasonaTE -mode annotate -projectFolder /path/to/dir -projectName $dirname -tool tirvish
reasonaTE -mode annotate -projectFolder /path/to/dir -projectName $dirname -tool transposonPSI
reasonaTE -mode annotate -projectFolder /path/to/dir -projectName $dirname -tool NCBICDD1000
```
As the programs like [RepeatModeler](https://github.com/Dfam-consortium/RepeatModeler), [RepeatMasker](https://github.com/rmhubley/RepeatMasker) & [ltrpred](https://github.com/HajkD/LTRpred) are not included within the package of *Transposon Annotator "reasonaTE"*, we need to install and run these programs separately.
```bash
# Running RepeatModeler
BuildDatabase -name sequence_index -engine ncbi $assembly
RepeatModeler -threads 100 -database sequence_index
```
With **RepeatMasker**, we can provide curated libraries of repeats during repeat library preparation. RepeatMasker, by default, considers repeat library from [Dfam](https://www.dfam.org/) databse.

```bash
# Retrieving repeat dabases
wget https://www.dfam.org/releases/Dfam_3.7/families/Dfam_curatedonly.h5.gz
gunzip Dfam_curatedonly.h5.gz

# Checking available repeats in Dfam database
python famdb.py -i /path/to/dir/Dfam_curatedonly.h5 names insecta

# Running RepeatMasker
assembly="/path/to/dir/draft.assembly.fa"
RepeatMasker -pa 50 $assembly -species insecta -gff
```
We can now check the completeness of each required analyses as desired by the *Transposon Annotator "reasonaTE"* pipeline.
```bash
reasonaTE -mode checkAnnotations -projectFolder /path/to/dir -projectName $dirname
```
