# Annotation
Annotation of repeat and CDS in the assembled nuclear genome

As an essential step, prior to the de-novo genome annotation, the repeat classes were identified and annotated in the genome. To prepare a comprehensive repeat library, a recently benchmarked pipeline that brings a brilliant compilation of analytical tools and publicly available databases under a common umbrella, **The Extensive *de novo* TE Annotator** [EDTA](https://github.com/oushujun/EDTA), was used. EDTA appears as an one stop solution for repeat library preparation. Alternatively, one can try [TransposonUltimate](https://github.com/DerKevinRiehl/TransposonUltimate) for library preparation. As a first step, the repeat classes which are known to be poorly represented in EDTA analysis were compiled from existing public repositories.

Compiling data from [SINEBase](https://sines.eimb.ru/).
```bash
# Retrieving insect specific SINE repeats
wget http://sines.eimb.ru/banks/SINEs.bnk
seqkit seq SINEs.bnk -n | grep "Insecta" > sines.insect.id

# Renaming SINE repeats as per EDTA requirement
seqkit grep -n -f sines.insect.id SINEs.bnk -o sines.insect.seq
cat sines.insect.seq | seqkit seq -i | seqkit replace -p $ -r _insecta_sinebase#SINE/Unknown > sines.insect.lib
```
Compiling data from [Dfam](https://www.dfam.org/home) database.
```bash
# Retrieving insect specific repeat classes from Dfam database
wget https://www.dfam.org/releases/Dfam_3.7/families/Dfam_curatedonly.h5.gz
gunzip Dfam_curatedonly.h5.gz
./famdb.py -i Dfam_curatedonly.h5 families -f fasta_name --include-class-in-name -d --class SINE 'Insecta' > dfam.insecta.sine.seq
./famdb.py -i Dfam_curatedonly.h5 families -f fasta_name --include-class-in-name -d --class LINE 'Insecta' > dfam.insecta.line.seq
./famdb.py -i Dfam_curatedonly.h5 families -f fasta_name --include-class-in-name -d --class Satellite 'Insecta' > dfam.insecta.sat.seq
./famdb.py -i Dfam_curatedonly.h5 families -f fasta_name --include-class-in-name -d --class rRNA 'Insecta' > dfam.insecta.rdna.seq

# Renaming the repeat classes as per EDTA requirement
cat dfam.insecta.sine.seq | seqkit seq -i | seqkit replace -p "\#.+" -r "_insecta_dfam#SINE/Unknown" > dfam.insecta.sine.lib
cat dfam.insecta.line.seq | seqkit seq -i | seqkit replace -p "\#.+" -r "_insecta_dfam#LINE/Unknown" > dfam.insecta.line.lib
cat dfam.insecta.sat.seq | seqkit seq -i | seqkit replace -p "\#.+" -r "_insecta_dfam#Satellite/Satellite" > dfam.insecta.sat.lib
cat dfam.insecta.rdna.seq | seqkit seq -i | seqkit replace -p "\#.+" -r "_insecta_dfam#rDNA/5S" > dfam.insecta.rdna.lib
```
Now, combining the compiled repeat classes.
```bash
cat sines.insect.lib dfam.insecta.sine.lib dfam.insecta.line.lib dfam.insecta.sat.lib dfam.insecta.rdna.lib > repeat.combined.lib
grep -c ">" repeat.combined.lib
```
To prevent errornous annotation of coding sequences as repeats, curated CDS regions from a related species was retrieved and supplied to EDTA during analysis.
```bash
# CDS data from LepDec
wget https://data.nal.usda.gov/system/files/lepdec_OGSv1-2.tar.gz (On 2023-07-19)
tar -xvzf lepdec_OGSv1-2.tar.gz
mv OGSv1.2/ lepdec.OGSv1.2
cp ../lepdec.OGSv1.2/lepdec_OGSv1.2_GCF_000500325.1_cds.fa ./lepdec.cds.fa
```
Repeat library preparation using [EDTA](https://github.com/oushujun/EDTA).
```bash
EDTA.pl --genome nuc.clean.fa --cds lepdec.cds.fa --curatedlib repeat.combined.lib --overwrite 1 --sensitive 1 --anno 1 --evaluate 0 --threads 100
cp ../EDTA/BP23_nuclear.fa.mod.EDTA.TElib.fa ./TElib.fa
```
> The ouput of EDTA analysis is a de-novo repeat library ```TElib.fa```.

Repeat annotation and genome soft-masking with [RepeatMasker](https://github.com/rmhubley/RepeatMasker) using the de-novo repeat library.
```bash
# Running repeatmasker
RepeatMasker -pa 25 nuc.clean.fa -lib TElib.fa -gff -xsmall
```
> As a result from this step, the nuclear genome is soft-masked for repeat regions ```nuc.mask.fa``` and ready for gene annotation.

```bash
###Retrieving protein databases
wget https://bioinf.uni-greifswald.de/bioinf/partitioned_odb11/Arthropoda.fa.gz
gunzip Arthropoda.fa.gz
wget https://data.nal.usda.gov/system/files/lepdec_OGSv1-2.tar.gz
tar -xvzf lepdec_OGSv1-2.tar.gz
cat  Arthropoda.fa ./lepdec.OGSv1.2/lepdec_OGSv1.2_GCF_000500325.1_pep.fa > proteins.fa
grep -c ">" proteins.fa      #4326149 seq

cat proteins.fa | seqkit replace -p "\s" -r "_" > proteins.edited.fa    #replacing whitespace in header
```

```bash
###Ab-inito Gene prediction with BRAKER3
cp ../masking/BP23_nuclear.fa.masked ./BP23_nuclear.masked.fa
cp ~/beetle/data/refseq/proteins.edited.fa .
braker.pl --genome=HYD23_masked.fa --prot_seq=proteins.edited.fa --threads 48 --gff3 --workingdir=bp23
```

```bash
#Annotation quality assessment
cp braker.codingseq ../../busco/brakerO.cds.fa
busco -i $aa --out_path /home/sahoork/beetle/analysis/annotation/busco -o braker-insecta -m protein -l $db1 -c 20 --offline --augustus_parameters='--progress=true' --download_path /home/sahoork/beetle/data/refseq/busco_downloads
busco -i $aa --out_path /home/sahoork/beetle/analysis/annotation/busco -o braker-endopt -m protein -l $db2 -c 20 --offline --augustus_parameters='--progress=true' --download_path /home/sahoork/beetle/data/refseq/busco_downloads
```

```bash
bedtools intersect -a braker.gff3 -b BP23_nuclear.fa.out.gff -wo > intersect.txt
cat intersect.txt | awk '{ if ($3 == "exon") print $0 }' | awk '{ sum += ($22) } END { print NR, sum}'
cat intersect.txt | awk '{ if ($3 == "intron") print $0 }' | awk '{ sum += ($22) } END { print NR, sum}'   # no of overlapping repeats, size of total overlap
cat intersect.txt | awk '{ if ($3 == "intron") print $9 }' | sort | uniq
```

